sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,382 @@
+                               config.chip_select.trail);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNLEAD_0_FIELD,
+                               config.chip_select.lead);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT,
+                             config.full_cycle);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPHA_0_BIT, config.cpha);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPOL_0_BIT, config.cpol);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_CONFIGOPTS_REG_OFFSET, reg);
+  spi_host_enable(spi_host, true);
+  return kDifOk;
+}
+
+static void wait_ready(const dif_spi_host_t *spi_host) {
+  bool ready;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    ready = bitfield_bit32_read(reg, SPI_HOST_STATUS_READY_BIT);
+  } while (!ready);
+}
+
+static void wait_tx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == SPI_HOST_PARAM_TX_DEPTH);
+}
+
+static void wait_rx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (rxqd == 0);
+}
+
+static inline void tx_fifo_write8(const dif_spi_host_t *spi_host,
+                                  uintptr_t srcaddr) {
+  uint8_t *src = (uint8_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
+                             uint16_t len) {
+  uintptr_t ptr = (uintptr_t)src;
+
+  // If the pointer starts mis-aligned, write until we are aligned.
+  while (misalignment32_of(ptr) && len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+
+  // Write complete 32-bit words to the fifo.
+  while (len > 3) {
+    tx_fifo_write32(spi_host, ptr);
+    ptr += 4;
+    len -= 4;
+  }
+
+  // Clean up any leftover bytes.
+  while (len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+}
+
+typedef struct queue {
+  int32_t length;
+  uint8_t alignas(uint64_t) data[8];
+} queue_t;
+
+static void enqueue_byte(queue_t *queue, uint8_t data) {
+  queue->data[queue->length++] = data;
+}
+
+static void enqueue_word(queue_t *queue, uint32_t data) {
+  if (queue->length % sizeof(uint32_t) == 0) {
```

Comment:
You could use `misalignment32_of` here.

Reply:
Although `misalignment32_of` will do the "right thing" here, it's not really what I mean.

The implementation of `misalignment32_of` checks to see if a pointer matches the architectural alignment constraints of a `uint32_t`, whereas I'm check to see if the current queue length is appropriate to receive a word.  If we happened to be on a CPU with a different alignment constraint for `uint32_t` (e.g. some CPU that permitted 1-byte alignment), then `misalignment32_of` wouldn't compute the right result.